File: examples/src/test/java/jdocs/cassandra/CassandraProjectionDocExample.java
/*
 * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.cassandra;

import java.time.Duration;

import akka.Done;
import akka.NotUsed;
import akka.actor.typed.ActorSystem;
import akka.actor.typed.javadsl.Behaviors;
import akka.stream.javadsl.FlowWithContext;
import jdocs.eventsourced.ShoppingCart;

//#daemon-imports
import akka.cluster.sharding.typed.javadsl.ShardedDaemonProcess;
import akka.cluster.sharding.typed.ShardedDaemonProcessSettings;
import akka.projection.ProjectionBehavior;

//#daemon-imports

//#source-provider-imports
import akka.persistence.cassandra.query.javadsl.CassandraReadJournal;
import akka.persistence.query.Offset;
import akka.projection.javadsl.SourceProvider;
import akka.projection.eventsourced.javadsl.EventSourcedProvider;
import akka.projection.eventsourced.EventEnvelope;

//#source-provider-imports

//#projection-imports
import akka.projection.cassandra.javadsl.CassandraProjection;
import akka.projection.Projection;
import akka.projection.ProjectionId;

//#projection-imports

//#handler-imports
import akka.projection.javadsl.Handler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

//#handler-imports

//#get-offset
import akka.projection.javadsl.ProjectionManagement;

//#get-offset

//#update-offset
import akka.persistence.query.Sequence;

//#update-offset

public interface CassandraProjectionDocExample {

  //#handler
  public class ShoppingCartHandler extends Handler<EventEnvelope<ShoppingCart.Event>> {
    private Logger logger = LoggerFactory.getLogger(getClass());

    @Override
    public CompletionStage<Done> process(EventEnvelope<ShoppingCart.Event> envelope) {
      ShoppingCart.Event event = envelope.event();
      if (event instanceof ShoppingCart.CheckedOut) {
        ShoppingCart.CheckedOut checkedOut = (ShoppingCart.CheckedOut) event;
        logger.info("Shopping cart {} was checked out at {}", checkedOut.cartId, checkedOut.eventTime);
        return CompletableFuture.completedFuture(Done.getInstance());
      } else {
        logger.debug("Shopping cart {} changed by {}", event.getCartId(), event);
        return CompletableFuture.completedFuture(Done.getInstance());
      }
    }
  }
  //#handler

  //#grouped-handler
  public class GroupedShoppingCartHandler extends Handler<List<EventEnvelope<ShoppingCart.Event>>> {
    private Logger logger = LoggerFactory.getLogger(getClass());

    @Override
    public CompletionStage<Done> process(List<EventEnvelope<ShoppingCart.Event>> envelopes) {
      envelopes.forEach(env -> {
        ShoppingCart.Event event = env.event();
        if (event instanceof ShoppingCart.CheckedOut) {
          ShoppingCart.CheckedOut checkedOut = (ShoppingCart.CheckedOut) event;
          logger.info("Shopping cart {} was checked out at {}", checkedOut.cartId, checkedOut.eventTime);
        } else {
          logger.debug("Shopping cart {} changed by {}", event.getCartId(), event);
        }
      });
      return CompletableFuture.completedFuture(Done.getInstance());
    }
  }
  //#grouped-handler

  public static void illustrateAtLeastOnce() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");

    //#sourceProvider
    SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider =
      EventSourcedProvider.eventsByTag(system, CassandraReadJournal.Identifier(), "carts-1");
    //#sourceProvider

    //#atLeastOnce
    int saveOffsetAfterEnvelopes = 100;
    Duration saveOffsetAfterDuration = Duration.ofMillis(500);

    Projection<EventEnvelope<ShoppingCart.Event>> projection =
      CassandraProjection.atLeastOnce(
        ProjectionId.of("shopping-carts", "carts-1"),
        sourceProvider,
        new ShoppingCartHandler()
      )
      .withSaveOffset(saveOffsetAfterEnvelopes, saveOffsetAfterDuration);
    //#atLeastOnce

  }

  public static void illustrateAtMostOnce() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");

    SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider =
      EventSourcedProvider.eventsByTag(system, CassandraReadJournal.Identifier(), "carts-1");

    //#atMostOnce
    Projection<EventEnvelope<ShoppingCart.Event>> projection =
      CassandraProjection.atMostOnce(
        ProjectionId.of("shopping-carts", "carts-1"),
        sourceProvider,
        new ShoppingCartHandler());
    //#atMostOnce

  }

  public static void illustrateGrouped() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");

    SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider =
      EventSourcedProvider.eventsByTag(system, CassandraReadJournal.Identifier(), "carts-1");

    //#grouped
    int groupAfterEnvelopes = 20;
    Duration groupAfterDuration = Duration.ofMillis(500);

    Projection<EventEnvelope<ShoppingCart.Event>> projection =
      CassandraProjection.groupedWithin(
        ProjectionId.of("shopping-carts", "carts-1"),
        sourceProvider,
        new GroupedShoppingCartHandler()
      )
      .withGroup(groupAfterEnvelopes, groupAfterDuration);
    //#grouped

  }

  public static void illustrateAtLeastOnceFlow() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");

    SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider =
      EventSourcedProvider.eventsByTag(system, CassandraReadJournal.Identifier(), "carts-1");

    //#atLeastOnceFlow

    Logger logger = LoggerFactory.getLogger("example");

    FlowWithContext<EventEnvelope<ShoppingCart.Event>, EventEnvelope<ShoppingCart.Event>, Done, EventEnvelope<ShoppingCart.Event>, NotUsed> flow =
      FlowWithContext.<EventEnvelope<ShoppingCart.Event>, EventEnvelope<ShoppingCart.Event>> create()
      .map(EventEnvelope::event)
      .map(event -> {
        if (event instanceof ShoppingCart.CheckedOut) {
          ShoppingCart.CheckedOut checkedOut = (ShoppingCart.CheckedOut) event;
          logger.info("Shopping cart {} was checked out at {}", checkedOut.cartId, checkedOut.eventTime);
        } else {
          logger.debug("Shopping cart {} changed by {}", event.getCartId(), event);
        }
        return Done.getInstance();
      });

    int saveOffsetAfterEnvelopes = 100;
    Duration saveOffsetAfterDuration = Duration.ofMillis(500);

    Projection<EventEnvelope<ShoppingCart.Event>> projection =
      CassandraProjection.atLeastOnceFlow(
        ProjectionId.of("shopping-carts", "carts-1"),
        sourceProvider,
        flow
      )
      .withSaveOffset(saveOffsetAfterEnvelopes, saveOffsetAfterDuration);
    //#atLeastOnceFlow

  }


  static class IllustrateRunningWithShardedDaemon {

    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");



    //#running-source-provider
    SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider(String tag) {
      return EventSourcedProvider.eventsByTag(system, CassandraReadJournal.Identifier(), tag);
    }
    //#running-source-provider

    //#running-projection
    int saveOffsetAfterEnvelopes = 100;
    Duration saveOffsetAfterDuration = Duration.ofMillis(500);

    Projection<EventEnvelope<ShoppingCart.Event>> projection(String tag) {
      return CassandraProjection.atLeastOnce(
        ProjectionId.of("shopping-carts", tag),
        sourceProvider(tag),
        new ShoppingCartHandler()
      )
      .withSaveOffset(saveOffsetAfterEnvelopes, saveOffsetAfterDuration);
    }
    //#running-projection


     public IllustrateRunningWithShardedDaemon() {

    //#running-with-daemon-process
     ShardedDaemonProcess.get(system).init(
       ProjectionBehavior.Command.class,
       "shopping-carts",
       ShoppingCart.tags.size(),
       id -> ProjectionBehavior.create(projection(ShoppingCart.tags.get(id))),
       ShardedDaemonProcessSettings.create(system),
       Optional.of(ProjectionBehavior.stopMessage())
     );
    //#running-with-daemon-process
     }
  }


  public static void illustrateProjectionSettings() {

    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");

    SourceProvider<Offset, EventEnvelope<ShoppingCart.Event>> sourceProvider =
            EventSourcedProvider.eventsByTag(system, CassandraReadJournal.Identifier(), "carts-1");

    //#projection-settings
    int saveOffsetAfterEnvelopes = 100;
    Duration saveOffsetAfterDuration = Duration.ofMillis(500);

    Projection<EventEnvelope<ShoppingCart.Event>> projection =
            CassandraProjection.atLeastOnce(
                ProjectionId.of("shopping-carts", "carts-1"),
                sourceProvider,
                new ShoppingCartHandler()
            ).withRestartBackoff(
                Duration.ofSeconds(10), /*minBackoff*/
                Duration.ofSeconds(60), /*maxBackoff*/
                0.5 /*randomFactor*/
            )
            .withSaveOffset(saveOffsetAfterEnvelopes, saveOffsetAfterDuration);
    //#projection-settings

  }

  public static void illustrateGetOffset() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");
    //#get-offset

    ProjectionId projectionId = ProjectionId.of("shopping-carts", "carts-1");
    CompletionStage<Optional<Offset>> currentOffset =
      ProjectionManagement.get(system).getOffset(projectionId);
    //#get-offset
  }

  public static void illustrateClearOffset() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");
    //#clear-offset

    ProjectionId projectionId = ProjectionId.of("shopping-carts", "carts-1");
    CompletionStage<Done> done = ProjectionManagement.get(system).clearOffset(projectionId);
    //#clear-offset
  }

  public static void illustrateUpdateOffset() {
    ActorSystem<Void> system = ActorSystem.create(Behaviors.empty(), "Example");
    //#update-offset

    ProjectionId projectionId = ProjectionId.of("shopping-carts", "carts-1");
    CompletionStage<Optional<Sequence>> currentOffset =
      ProjectionManagement.get(system).getOffset(projectionId);
    currentOffset.thenAccept(optionalOffset -> {
      if (optionalOffset.isPresent()) {
        Sequence newOffset = new Sequence(optionalOffset.get().value());
        CompletionStage<Done> done = ProjectionManagement.get(system).updateOffset(projectionId, newOffset);
      }
    });
    //#update-offset
  }
}

File: examples/src/it/scala/docs/cassandra/CassandraProjectionDocExample.scala
/*
 * Copyright (C) 2020-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.cassandra

import akka.actor.typed.scaladsl.LoggerOps
import akka.actor.typed.ActorSystem
import akka.actor.typed.scaladsl.Behaviors
import akka.projection.ProjectionContext
import akka.projection.eventsourced.EventEnvelope
import akka.stream.scaladsl.FlowWithContext
//#daemon-imports
import akka.cluster.sharding.typed.scaladsl.ShardedDaemonProcess
import akka.projection.ProjectionBehavior

//#daemon-imports

//#source-provider-imports
import akka.persistence.cassandra.query.scaladsl.CassandraReadJournal
import akka.projection.eventsourced.scaladsl.EventSourcedProvider
import docs.eventsourced.ShoppingCart

//#source-provider-imports

//#projection-imports
import akka.projection.ProjectionId
import akka.projection.cassandra.scaladsl.CassandraProjection

//#projection-imports

//#handler-imports
import scala.concurrent.duration._
import scala.concurrent.Future

import akka.Done
import akka.projection.scaladsl.Handler
import org.slf4j.LoggerFactory

//#handler-imports

object CassandraProjectionDocExample {

  private val system = ActorSystem[Nothing](Behaviors.empty, "Example")

  //#handler
  class ShoppingCartHandler extends Handler[EventEnvelope[ShoppingCart.Event]] {
    private val logger = LoggerFactory.getLogger(getClass)

    override def process(envelope: EventEnvelope[ShoppingCart.Event]): Future[Done] = {
      envelope.event match {
        case ShoppingCart.CheckedOut(cartId, time) =>
          logger.info2("Shopping cart {} was checked out at {}", cartId, time)
          Future.successful(Done)

        case otherEvent =>
          logger.debug2("Shopping cart {} changed by {}", otherEvent.cartId, otherEvent)
          Future.successful(Done)
      }
    }
  }
  //#handler

  //#grouped-handler
  import scala.collection.immutable

  class GroupedShoppingCartHandler extends Handler[immutable.Seq[EventEnvelope[ShoppingCart.Event]]] {
    private val logger = LoggerFactory.getLogger(getClass)

    override def process(envelopes: immutable.Seq[EventEnvelope[ShoppingCart.Event]]): Future[Done] = {
      envelopes.map(_.event).foreach {
        case ShoppingCart.CheckedOut(cartId, time) =>
          logger.info2("Shopping cart {} was checked out at {}", cartId, time)

        case otherEvent =>
          logger.debug2("Shopping cart {} changed by {}", otherEvent.cartId, otherEvent)
      }
      Future.successful(Done)
    }
  }
  //#grouped-handler

  //#sourceProvider
  val sourceProvider =
    EventSourcedProvider
      .eventsByTag[ShoppingCart.Event](system, readJournalPluginId = CassandraReadJournal.Identifier, tag = "carts-1")
  //#sourceProvider

  object IllustrateAtLeastOnce {
    //#atLeastOnce
    val projection =
      CassandraProjection
        .atLeastOnce(
          projectionId = ProjectionId("shopping-carts", "carts-1"),
          sourceProvider,
          handler = () => new ShoppingCartHandler)
        .withSaveOffset(afterEnvelopes = 100, afterDuration = 500.millis)
    //#atLeastOnce
  }

  object IllustrateAtMostOnce {
    //#atMostOnce
    val projection =
      CassandraProjection.atMostOnce(
        projectionId = ProjectionId("shopping-carts", "carts-1"),
        sourceProvider,
        handler = () => new ShoppingCartHandler)
    //#atMostOnce
  }

  object IllustrateGrouped {
    //#grouped
    val projection =
      CassandraProjection
        .groupedWithin(
          projectionId = ProjectionId("shopping-carts", "carts-1"),
          sourceProvider,
          handler = () => new GroupedShoppingCartHandler)
        .withGroup(groupAfterEnvelopes = 20, groupAfterDuration = 500.millis)
    //#grouped
  }

  object IllustrateAtLeastOnceFlow {
    //#atLeastOnceFlow
    val logger = LoggerFactory.getLogger(getClass)

    val flow = FlowWithContext[EventEnvelope[ShoppingCart.Event], ProjectionContext]
      .map(envelope => envelope.event)
      .map {
        case ShoppingCart.CheckedOut(cartId, time) =>
          logger.info2("Shopping cart {} was checked out at {}", cartId, time)
          Done

        case otherEvent =>
          logger.debug2("Shopping cart {} changed by {}", otherEvent.cartId, otherEvent)
          Done
      }

    val projection =
      CassandraProjection
        .atLeastOnceFlow(projectionId = ProjectionId("shopping-carts", "carts-1"), sourceProvider, handler = flow)
        .withSaveOffset(afterEnvelopes = 100, afterDuration = 500.millis)
    //#atLeastOnceFlow
  }

  object IllustrateRecoveryStrategy {
    //#withRecoveryStrategy
    import akka.projection.HandlerRecoveryStrategy

    val projection =
      CassandraProjection
        .atLeastOnce(
          projectionId = ProjectionId("shopping-carts", "carts-1"),
          sourceProvider,
          handler = () => new ShoppingCartHandler)
        .withRecoveryStrategy(HandlerRecoveryStrategy.retryAndFail(retries = 10, delay = 1.second))
    //#withRecoveryStrategy
  }

  object IllustrateRestart {
    //#withRestartBackoff
    val projection =
      CassandraProjection
        .atLeastOnce(
          projectionId = ProjectionId("shopping-carts", "carts-1"),
          sourceProvider,
          handler = () => new ShoppingCartHandler)
        .withRestartBackoff(minBackoff = 200.millis, maxBackoff = 5.seconds, randomFactor = 0.1)
    //#withRestartBackoff
  }

  object IllustrateRunningWithShardedDaemon {

    //#running-source-provider
    def sourceProvider(tag: String) =
      EventSourcedProvider
        .eventsByTag[ShoppingCart.Event](
          system = system,
          readJournalPluginId = CassandraReadJournal.Identifier,
          tag = tag)
    //#running-source-provider

    //#running-projection
    def projection(tag: String) =
      CassandraProjection
        .atLeastOnce(
          projectionId = ProjectionId("shopping-carts", tag),
          sourceProvider(tag),
          handler = () => new ShoppingCartHandler)
        .withSaveOffset(100, 500.millis)
    //#running-projection

    //#running-with-daemon-process
    ShardedDaemonProcess(system).init[ProjectionBehavior.Command](
      name = "shopping-carts",
      numberOfInstances = ShoppingCart.tags.size,
      behaviorFactory = (i: Int) => ProjectionBehavior(projection(ShoppingCart.tags(i))),
      stopMessage = ProjectionBehavior.Stop)
    //#running-with-daemon-process
  }

  object IllustrateRunningWithActor {

    Behaviors.setup[String] { context =>
      //#running-with-actor
      def sourceProvider(tag: String) =
        EventSourcedProvider
          .eventsByTag[ShoppingCart.Event](
            system = system,
            readJournalPluginId = CassandraReadJournal.Identifier,
            tag = tag)

      def projection(tag: String) =
        CassandraProjection
          .atLeastOnce(
            projectionId = ProjectionId("shopping-carts", tag),
            sourceProvider(tag),
            handler = () => new ShoppingCartHandler)

      val projection1 = projection("carts-1")

      context.spawn(ProjectionBehavior(projection1), projection1.projectionId.id)
      //#running-with-actor

      Behaviors.empty
    }
  }

  object IllustrateRunningWithSingleton {

    //#running-with-singleton
    import akka.cluster.typed.ClusterSingleton
    import akka.cluster.typed.SingletonActor

    def sourceProvider(tag: String) =
      EventSourcedProvider
        .eventsByTag[ShoppingCart.Event](
          system = system,
          readJournalPluginId = CassandraReadJournal.Identifier,
          tag = tag)

    def projection(tag: String) =
      CassandraProjection
        .atLeastOnce(
          projectionId = ProjectionId("shopping-carts", tag),
          sourceProvider(tag),
          handler = () => new ShoppingCartHandler)

    val projection1 = projection("carts-1")

    ClusterSingleton(system).init(SingletonActor(ProjectionBehavior(projection1), projection1.projectionId.id))
    //#running-with-singleton

  }

  object IllustrateProjectionSettings {

    //#projection-settings
    val projection =
      CassandraProjection
        .atLeastOnce(
          projectionId = ProjectionId("shopping-carts", "carts-1"),
          sourceProvider,
          handler = () => new ShoppingCartHandler)
        .withRestartBackoff(minBackoff = 10.seconds, maxBackoff = 60.seconds, randomFactor = 0.5)
        .withSaveOffset(100, 500.millis)
    //#projection-settings

  }

  object IllustrateGetOffset {
    //#get-offset
    import akka.projection.scaladsl.ProjectionManagement
    import akka.persistence.query.Offset
    import akka.projection.ProjectionId

    val projectionId = ProjectionId("shopping-carts", "carts-1")
    val currentOffset: Future[Option[Offset]] = ProjectionManagement(system).getOffset[Offset](projectionId)
    //#get-offset
  }

  object IllustrateClearOffset {
    import akka.projection.scaladsl.ProjectionManagement
    //#clear-offset
    val projectionId = ProjectionId("shopping-carts", "carts-1")
    val done: Future[Done] = ProjectionManagement(system).clearOffset(projectionId)
    //#clear-offset
  }

  object IllustrateUpdateOffset {
    import akka.projection.scaladsl.ProjectionManagement
    import system.executionContext
    //#update-offset
    import akka.persistence.query.Sequence
    val projectionId = ProjectionId("shopping-carts", "carts-1")
    val currentOffset: Future[Option[Sequence]] = ProjectionManagement(system).getOffset[Sequence](projectionId)
    currentOffset.foreach {
      case Some(s) => ProjectionManagement(system).updateOffset[Sequence](projectionId, Sequence(s.value + 1))
      case None    => // already removed
    }
    //#update-offset
  }

  object IllustratePauseResume {
    import akka.projection.scaladsl.ProjectionManagement
    import system.executionContext
    def someDataMigration() = Future.successful(Done)
    //#pause-resume
    import akka.projection.scaladsl.ProjectionManagement
    import akka.projection.ProjectionId

    val projectionId = ProjectionId("shopping-carts", "carts-1")
    val mgmt = ProjectionManagement(system)
    val done = for {
      _: Future[Done] <- mgmt.pauseProjection(projectionId)
      _ <- someDataMigration()
      _: Future[Done] <- mgmt.resumeProjection(projectionId)
    } yield Done
    //#pause-resume
  }

  object IllustrateIsPaused {
    import akka.projection.scaladsl.ProjectionManagement
    import akka.projection.ProjectionId
    //#is-paused
    val projectionId = ProjectionId("shopping-carts", "carts-1")
    val paused: Future[Boolean] = ProjectionManagement(system).isProjectionPaused(projectionId)
    //#is-paused
  }

}
